File: backend/app/api/v1/checkin.py
"""
打卡接口 - 打卡、连续天数统计、日历；Sprint 3.3
"""
from datetime import date
from fastapi import APIRouter, Depends, Query

from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.user import User
from app.schemas.checkin import CheckInCreate, CheckInCalendarResponse, CheckInStatsResponse
from app.services.checkin_service import (
    check_in,
    get_checkin_calendar,
    get_checkin_stats,
)
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter(prefix="/checkin", tags=["checkin"])


@router.post("", response_model=dict)
async def create_checkin(
    body: CheckInCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """打卡"""
    checkin = await check_in(db, current_user.id, body.check_date, body.note)
    return {
        "id": checkin.id,
        "check_date": checkin.check_date.isoformat(),
        "note": checkin.note,
    }


@router.get("/calendar")
async def checkin_calendar(
    year: int = Query(..., ge=2000, le=2100),
    month: int = Query(..., ge=1, le=12),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """获取某月的打卡日历"""
    items = await get_checkin_calendar(db, current_user.id, year, month)
    return CheckInCalendarResponse(items=items)


@router.get("/stats")
async def checkin_stats(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """获取用户打卡统计：总天数、本月天数、当前连续天数"""
    stats = await get_checkin_stats(db, current_user.id)
    return CheckInStatsResponse(
        total_days=stats["total_days"],
        this_month=stats["this_month"],
        current_streak=stats["current_streak"],
    )

File: miniapp/src/api/checkin.ts
/**
 * 签到 - 与 backend /api/v1/checkin 一致
 */
import request from '@/utils/request'

const PREFIX = '/api/v1/checkin'

/** 打卡创建请求 */
export interface CheckInCreateReq {
  check_date: string // ISO date string
  note?: string
}

/** 打卡响应 */
export interface CheckInRes {
  id: string
  check_date: string
  note: string | null
}

/** 打卡日历项 */
export interface CalendarItem {
  date: string
  checked: boolean
  note?: string
}

/** 打卡日历响应 */
export interface CalendarRes {
  items: CalendarItem[]
}

/** 打卡统计响应 */
export interface CheckInStatsRes {
  total_days: number
  this_month: number
  current_streak: number
}

/** 创建打卡 */
export function createCheckIn(data: CheckInCreateReq) {
  return request<CheckInRes>({
    url: PREFIX,
    method: 'POST',
    data,
  })
}

/** 获取打卡日历 */
export function getCheckInCalendar(year: number, month: number) {
  return request<CalendarRes>({
    url: `${PREFIX}/calendar?year=${year}&month=${month}`,
    method: 'GET',
  })
}

/** 获取打卡统计 */
export function getCheckInStats() {
  return request<CheckInStatsRes>({
    url: `${PREFIX}/stats`,
    method: 'GET',
  })
}

/** 打卡列表项（用于用户主页展示） */
export interface CheckinItem {
  id: string
  check_date: string
  note: string | null
}

/** 打卡列表响应 */
export interface CheckinListRes {
  items: CheckinItem[]
  total: number
}

/** 获取打卡列表（分页） */
export function getCheckinList(params: { limit?: number; offset?: number }): Promise<CheckinListRes> {
  const { limit = 30, offset = 0 } = params
  return request<CheckinListRes>({
    url: `${PREFIX}/list?limit=${limit}&offset=${offset}`,
    method: 'GET',
  })
}
